subscriptions/subscriptions-impl/src/main/java/com/duckduckgo/subscriptions/impl/SubscriptionFeaturesFetcher.kt

```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright (c) 2024 DuckDuckGo
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.duckduckgo.subscriptions.impl
+
+import androidx.lifecycle.LifecycleOwner
+import com.duckduckgo.app.di.AppCoroutineScope
+import com.duckduckgo.app.lifecycle.MainProcessLifecycleObserver
+import com.duckduckgo.di.scopes.AppScope
+import com.duckduckgo.subscriptions.impl.SubscriptionsConstants.BASIC_SUBSCRIPTION
+import com.duckduckgo.subscriptions.impl.billing.PlayBillingManager
+import com.duckduckgo.subscriptions.impl.repository.AuthRepository
+import com.duckduckgo.subscriptions.impl.services.SubscriptionsService
+import com.squareup.anvil.annotations.ContributesMultibinding
+import javax.inject.Inject
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.flow.first
+import kotlinx.coroutines.launch
+import logcat.logcat
+
+@ContributesMultibinding(
+    scope = AppScope::class,
+    boundType = MainProcessLifecycleObserver::class,
+)
+class SubscriptionFeaturesFetcher @Inject constructor(
+    @AppCoroutineScope private val appCoroutineScope: CoroutineScope,
+    private val playBillingManager: PlayBillingManager,
+    private val subscriptionsService: SubscriptionsService,
+    private val authRepository: AuthRepository,
+    private val privacyProFeature: PrivacyProFeature,
+) : MainProcessLifecycleObserver {
+
+    override fun onCreate(owner: LifecycleOwner) {
+        super.onCreate(owner)
+        appCoroutineScope.launch {
+            try {
+                if (privacyProFeature.featuresApi().isEnabled()) {
+                    fetchSubscriptionFeatures()
+                }
+            } catch (e: Exception) {
+                logcat { "Failed to fetch subscription features" }
+            }
+        }
+    }
+
+    private suspend fun fetchSubscriptionFeatures() {
+        playBillingManager.productsFlow
```

Comment:
Should we set this to happen in io?

Reply:
Default dispatcher is fine here - `playBillingManager`, `authRepository`, `subscriptionsService` should all handle switching to correct dispatcher internally when needed, so their APIs never block and are safe to use using any dispatcher.

That said, I just noticed that there are a couple of missing dispatcher declarations inside `authRepository`, will add a commit that fixes it.